android-components/components/feature/fxsuggest/src/main/java/mozilla/components/feature/fxsuggest/GlobalFxSuggestDependencyProvider.kt

```diff
@@ -0,0 +1,42 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.fxsuggest
+
+import mozilla.appservices.suggest.SuggestStore
+
+/**
+ * Provides global access to the dependencies needed to access Firefox Suggest search suggestions.
+ */
+object GlobalFxSuggestDependencyProvider {
+    /**
+     * The default Suggest database file name.
+     *
+     * Your application should pass this constant to [android.content.Context.getDatabasePath],
+     * then pass the resulting absolute path to
+     * [GlobalFxSuggestDependencyProvider.openSuggestStore].
+     */
+    const val DEFAULT_SUGGEST_DATABASE_NAME = "suggest.sqlite"
+
+    internal var suggestStore: SuggestStore? = null
+
+    /**
+     * Opens a Suggest store that persists suggestions, icons, and metadata in a database at the
+     * given path.
+     *
+     * Your application's [onCreate][android.app.Application.onCreate] method should call this
+     * method once.
+     *
+     * @param databasePath The absolute path of the Suggest database.
+     */
+    fun openSuggestStore(databasePath: String) {
```

Comment:
In your [WIP fenix patch](https://github.com/linabutler/firefox-android/commit/bf5747c7359588943a6a6acb1e1206ea75af62d6#diff-c07dbf42136256cfd1eea0fe95204636f9327f15255ec1fab0843e18bd2aaec7R309-R311), we suppress a strict mode violation which uses an IO API working on the main thread.

We may need either:
1. Moving the entire dependency provider initialization to happen on an IO dispatcher, and silently catch the `requireNotNull` exception that we may throw below (and this seems odd as I write it out)
2. Waiting until initialization is complete, and only then adding the AwesomeBar provider.

Reply:
To add a bit of detail to the `2` option, I would suggest:
1. make this function suspending 
2. convert this to a class that receives a `Context`
3. initialize the `suggestStore` as a lazy deferred property
```
class GlobalFxSuggestDependencyProvider(
    context: Context, // val isn't required here, and will mean the class drops the reference to the god object after the property is computed
    private val coroutineScope = CoroutineScope(Dispatchers.IO),
) {

    private val suggestStore: Deferred<SuggestStore> = coroutineScope.async(start = CoroutineStart.Lazy) {
        SuggestStore(context.getDatabasePath(...)
    }

    suspend fun requireSuggestStore(): SuggestStore {
        suggestStore.await()
    }
}
```

It seems like every usage of the store in your WIP patch is either already in a suspending context, or we can simply launch a new coroutine and handle the result asynchronously as Jonathan is suggesting.

Reply:
Ooh, I like that approach, @MatthewTighe! It's a little unfortunate that we try to avoid main-thread I/O in `SuggestStore` by having it open the database connections lazily, but Android's `Context.getDatabasePath` still ends up doing I/O to ensure the databases directory exists 😞 Oh, well!

Reply:
@MatthewTighe Hmm, I just noticed that your suggestion makes `GlobalFxSuggestDependencyProvider` a class. Where would we store an instance of that class, so that both the background worker and suggestion provider can access it?

It looks like Fenix's `FenixApplication.components` would be a good place—except the worker and suggestion provider are in Android Components code, so I don't think we have a concrete application context here that we can use to store additional fields, right?

I wonder if we can work around this with double indirection, though: keep `GlobalFxSuggestDependencyProvider` an object, as now, but have it hold a reference to a `FxSuggestLazyStore` wrapper that does the lazy initialization of the `SuggestStore`.

Something like:

```kt
object GlobalFxSuggestDependencyProvider {
    internal var lazyStore: FxSuggestLazyStore? = null

    fun initialize(lazyStore: FxSuggestLazyStore) {
        self.lazyStore = lazyStore
    }

    fun requireLazyStore(): FxSuggestLazyStore {
        // ...
    }
}

class FxSuggestLazyStore(
    context: Context,
    private val coroutineScope = CoroutineScope(Dispatchers.IO),
) {
    private val store: Deferred<SuggestStore> = coroutineScope.async(start = CoroutineStart.Lazy) {
        SuggestStore(context.getDatabasePath(...))
    }

    suspend fun get(): SuggestStore = store.await()
}
```

...And then Fenix would do something like:

```kt
val lazyStore = FxSuggestLazyStore(context)
GlobalFxSuggestDependencyProvider.initialize(lazyStore)
```

WDYT? Is there a simpler way? 😅

Reply:
I pushed 4378d42468765b3a624070628f8ef085dba7790d to show how this might look; feedback welcome! The new `FxSuggestStorage` class (what I called `FxSuggestLazyStore` in my comment above) also manages separate coroutine scopes for reading and writing.